src/cmd/build.rs

```diff
@@ -249,10 +257,74 @@ fn post_process_wasm(crate_metadata: &CrateMetadata) -> Result<()> {
     ensure_maximum_memory_pages(&mut module, MAX_MEMORY_PAGES)?;
     strip_custom_sections(&mut module);
 
+    validate_import_section(&module)?;
+
     parity_wasm::serialize_to_file(&crate_metadata.dest_wasm, module)?;
     Ok(())
 }
 
+/// Validates the import section in the Wasm.
+///
+/// The checks currently fall into two categories:
+/// - Known bugs for which we want to recommend a solution.
+/// - Markers inserted by the ink! codegen for errors which can't be checked at compile time.
+fn validate_import_section(module: &Module) -> Result<()> {
```

Comment:
I guess the main thing to discuss is if we want to have this validation in a more structured manner (e.g. as a separate module with separate functions for each validation criteria). For now there are only these three checks, but we already have some ideas of stuff to add.

Reply:
Generally the scheme is to have all of ink!'s generated errors have the same form which `cargo-contract` can then query for.
Currently the form is: `__ink_enforce_error_for_{0}_{1}`
Where `{0}` is the subject, which currently is only either message or construct but which will potentially have more subjects to come in the future and the `{1}` is additional context for the error message.
Ideally we would tune the additional error context information for our current errors to encompass much more information than what they currently do - which is just the selector of the message. Ideally the additional error context was better structured like so:
`__ink_enforced_error_{0}_{2..n}`
Where `{0}` still is an error code known to `cargo-contract`. Every error code required an additional `N` number of "parameters" that are given by `_` separated additional values. Then `cargo-contract` could construct a fully qualified error message to the user.
An example for this was:
```rust
// The below error code represents calling a `&mut self` message in a context that
// only allows for `&self` messages. This may happen under certain circumstances
// when ink! trait implementations are involved with long-hand calling notation.
//
// The error code requires 2 parameters:
// 1. The name of the called message.
// 2. The selector of the called message.
// 3. The trait that defines the called message.
const CALL_MUT_MESSAGE_INVALID_CTX_ERRC: u32 = 0x1;

fn __ink_enforced_error_1_foo_0xc0decafe_FooTrait
```
Then `1` refers to `CALL_MUT_MESSAGE_INVALID_CTX_ERRC`, `foo` is the message's name, `0xc0decafe` is the message's selector and `FooTrait` is the name of the trait that defines the function.
What we need to do is to articulate a proper list between ink! and `cargo-contract` for all the valid error codes and the parameters that `cargo-contract` can expect for them.

Reply:
That's great.

>What we need to do is to articulate a proper list between ink! and cargo-contract for all the valid error codes and the parameters that cargo-contract can expect for them.

Hmm, how about creating a crate in ink! for this and sharing it as a dependency?

Reply:
Perhaps it could go in the existing `metadata` crate? It could be a separate "static" metadata module.

Reply:
> That's great.
> 
> > What we need to do is to articulate a proper list between ink! and cargo-contract for all the valid error codes and the parameters that cargo-contract can expect for them.
> 
> Hmm, how about creating a crate in ink! for this and sharing it as a dependency?

How do you imagine this to work out exactly?
How are the structures encoding the errors, their required parameters and their display messages?

Reply:
> Perhaps it could go in the existing `metadata` crate? It could be a separate "static" metadata module.

Not sure metadata fits the use case. Can you elaborate?

Reply:
Just thinking out loud really, I initially thought of it because `metadata` is a crate which is already shared by `cargo-contract`. It does already contain some information about the environment so not too much of a stretch to imagine adding these error mappings.

Reply:
> I guess the main thing to discuss is if we want to have this validation in a more structured manner (e.g. as a separate module with separate functions for each validation criteria). For now there are only these three checks, but we already have some ideas for stuff to add.

If we know we will add more checks then I think we should just have them defined in a separate module at least to avoid cluttering up this file.